src/Ace/Halogen/Component.purs

```diff
@@ -171,94 +171,97 @@ initialAceState =
 
 -- | The Ace component.
 aceComponent
-  :: forall eff g
+  ∷ ∀ eff g
    . (Monad g, Affable (AceEffects eff) g)
-  => (Editor -> g Unit)
-  -> Maybe Autocomplete
-  -> H.Component AceState AceQuery g
+  ⇒ (Editor → g Unit)
+  → Maybe Autocomplete
+  → H.Component AceState AceQuery g
 aceComponent setup resume = H.lifecycleComponent
     { render
     , eval: eval setup resume
     , initializer: Just (H.action Init)
     , finalizer: Just (H.action Quit)
     }
 
-render :: AceState -> H.ComponentHTML AceQuery
+render ∷ AceState → H.ComponentHTML AceQuery
 render = const $ HH.div [ HP.ref (H.action <<< SetElement) ] []
 
-eval :: forall eff g
+eval ∷ ∀ eff g
  . (Monad g, Affable (AceEffects eff) g)
- => (Editor -> g Unit)
- -> Maybe Autocomplete
- -> AceQuery
+ ⇒ (Editor → g Unit)
+ → Maybe Autocomplete
+ → AceQuery
  ~> H.ComponentDSL AceState AceQuery g
 eval setup resume = case _ of
-  SetElement el next ->
-    H.modify (_ { element = el }) $> next
-
-  Init next -> do
-    el <- H.gets _.element
-    case el of
-      Nothing -> pure unit
-      Just el' -> do
-        key <- H.gets _.key >>= maybe (H.fromEff genKey) pure
-        editor <- H.fromEff $ Ace.editNode el' Ace.ace
-        H.set { key: Just key, editor: Just editor, element: Just el' }
-        H.fromEff do
-          enableAutocomplete
-          setAutocompleteResume resume editor
-          Editor.onFocus editor $ writeRef focused key
-        session <- H.fromEff $ Editor.getSession editor
-        H.subscribe $ H.eventSource_ (Session.onChange session) do
-          pure $ H.action TextChanged
-        H.liftH $ setup editor
+  SetElement el next → do
+    state ← H.get
+    H.modify _{ element = el }
+    for_ state.editor \editor → do
+      H.fromEff $ Editor.destroy editor
+      H.modify _{ editor = Nothing }
```

Comment:
I'd suggest combining the `modify` calls here to avoid multiple renders. Something like:

``` purescript
SetElement el next → do
  state ← H.get
  for_ state.editor (H.fromEff <<< Editor.destroy)
  H.modify _{ element = el, editor = Nothing }
```